src/libraries/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs

```diff
@@ -360,7 +360,9 @@ public Guid ClientConnectionId
                 else
                 {
                     Task reconnectTask = _currentReconnectionTask;
-                    if (reconnectTask != null && !reconnectTask.IsCompleted)
+                    // Connection closed but previously open should return the correct ClientConnectionId
+                    DbConnectionClosedPreviouslyOpened innerConnectionClosed = (InnerConnection as DbConnectionClosedPreviouslyOpened);
+                    if ((reconnectTask != null && !reconnectTask.IsCompleted) || null != innerConnectionClosed)
```

Comment:
Wouldn't this be better as
```C#
if ((reconnectTask != null && !reconnectTask.IsCompleted) || InnerConnection is DbConnectionClosedPreviouslyOpened)
{
   ...
}
```
?  That way we're not doing the cast unless we have to, we're not adding an unnecessary local, etc.